tests/unit/onyxClearNativeStorageTest.js

```diff
@@ -94,15 +92,70 @@ describe('Set data while storage is clearing', () => {
         Onyx.clear();
         return waitForPromisesToResolve()
             .then(() => {
-                // Then the value in Onyx and the cache is the default key state
+                // Then the value in Onyx, the cache, and Storage is the default key state
                 expect(onyxValue).toBe(DEFAULT_VALUE);
                 const cachedValue = cache.getValue(ONYX_KEYS.DEFAULT_KEY);
                 expect(cachedValue).toBe(DEFAULT_VALUE);
-                const storedValue = Storage.getItem(ONYX_KEYS.DEFAULT_KEY);
+                return Storage.getItem(ONYX_KEYS.DEFAULT_KEY)
+                    .then(storedValue => expect(parseInt(storedValue, 10)).toBe(DEFAULT_VALUE));
+            });
+    });
+
+    it('should preserve the value of any keysToPreserve passed in', () => {
+        expect.assertions(6);
+
+        // Given that Onyx has a value, and we have a variable listening to that value
+        let valueToKeep;
+        Onyx.connect({
+            key: ONYX_KEYS.REGULAR_KEY,
+            initWithStoredValues: false,
+            callback: val => valueToKeep = val,
+        });
+        Onyx.set(ONYX_KEYS.REGULAR_KEY, SET_VALUE);
+
+        // When clear is called with a key to preserve
+        Onyx.clear([ONYX_KEYS.REGULAR_KEY]);
+        return waitForPromisesToResolve()
+            .then(() => {
+                // Then the value in Onyx and the cache for the default key is the default key state
+                expect(onyxValue).toBe(DEFAULT_VALUE);
+                const defaultKeyCachedValue = cache.getValue(ONYX_KEYS.DEFAULT_KEY);
+                expect(defaultKeyCachedValue).toBe(DEFAULT_VALUE);
+                const defaultKeyStoredValue = Storage.getItem(ONYX_KEYS.DEFAULT_KEY);
 
                 // Then the value in Storage is null
-                // The default key state is never stored during Onyx.clear
-                return expect(storedValue).resolves.toBeNull();
+                expect(defaultKeyStoredValue).resolves.toBeNull();
+
+                // Then the value of the preserved key is also still set in both the cache and storage
+                expect(valueToKeep).toBe(SET_VALUE);
+                const regularKeyCachedValue = cache.getValue(ONYX_KEYS.REGULAR_KEY);
+                expect(regularKeyCachedValue).toBe(SET_VALUE);
+                return Storage.getItem(ONYX_KEYS.REGULAR_KEY)
+                    .then(storedValue => expect(parseInt(storedValue, 10)).toBe(SET_VALUE));
+            });
+    });
+
+    it('should preserve the value of any keysToPreserve over any default key states', () => {
+        expect.assertions(3);
+
+        // Given that Onyx has a value for a key with a default, and we have a variable listening to that value
+        Onyx.connect({
+            key: ONYX_KEYS.DEFAULT_KEY,
+            initWithStoredValues: false,
+            callback: val => onyxValue = val,
+        });
```

Comment:
We already set up this Onyx connection in beforeEach, so I don't think we need this. Do you agree?

Reply:
Yes - although maybe we should rename it to `defaultKeyOnyxValue`?

Reply:
Sure, sounds good to me.

Reply:
Actually, I ended up calling the variable in the test above `regularKeyOnyxValue`, which I think is a clear enough distinction for now.